wcfsetup/install/files/lib/system/background/job/UnfurlUrlBackgroundJob.class.php

```diff
@@ -0,0 +1,275 @@
+    {
+        switch ($this->getFailures()) {
+            case 1:
+                // 5 minutes
+                return 5 * 60;
+            case 2:
+                // 30 minutes
+                return 30 * 60;
+            case 3:
+                // 2 hours
+                return 2 * 60 * 60;
+        }
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function perform()
+    {
+        $unfurlUrl = new UnfurlUrl($this->urlID);
+
+        try {
+            $unfurlResponse = UnfurlResponse::fetch($unfurlUrl->url);
+
+            if (empty($unfurlResponse->getTitle())) {
+                $this->save(UnfurlUrl::STATUS_REJECTED);
+
+                return;
+            }
+
+            $title = StringUtil::truncate($unfurlResponse->getTitle(), 255);
+            $description = "";
+            if ($unfurlResponse->getDescription()) {
+                $description = StringUtil::truncate($unfurlResponse->getDescription(), 160);
+            }
+
+            $imageData = [];
+            $imageID = null;
+            $imageUrl = $unfurlResponse->getImageUrl();
+            if ($imageUrl) {
+                if (
+                    \strpos($imageUrl, '\\') === false
+                    && \strpos($imageUrl, "'") === false
+                    && Url::is($imageUrl)
+                ) {
+                    $imageID = self::getImageIdByUrl($imageUrl);
+
+                    if ($imageID === null) {
+                        $imageData = $this->getImageData($unfurlResponse);
+                    }
+                }
+            }
+
+            $this->save(
+                UnfurlUrl::STATUS_SUCCESSFUL,
+                $title,
+                $description,
+                $imageID,
+                $imageData
+            );
+        } catch (UrlInaccessible | ParsingFailed $e) {
+            if (\ENABLE_DEBUG_MODE) {
+                \wcf\functions\exception\logThrowable($e);
+            }
+
+            $this->save(UnfurlUrl::STATUS_REJECTED);
+        }
+    }
+
+    private function getImageData(UnfurlResponse $unfurlResponse): array
+    {
+        if (empty($unfurlResponse->getImageUrl()) || !Url::is($unfurlResponse->getImageUrl())) {
+            throw new \InvalidArgumentException("The given response does not have an image.");
+        }
+
+        try {
+            $imageResponse = $unfurlResponse->getImage();
+            $image = $this->downloadImage($imageResponse);
+            $imageData = \getimagesizefromstring($image);
+
+            if ($imageData === false) {
+                return [];
+            }
+
+            if (!$this->validateImage($imageData)) {
+                return [];
+            }
+
+            $imageSaveData = [];
+            $imageSaveData['imageUrl'] = $unfurlResponse->getImageUrl();
+            $imageSaveData['imageUrlHash'] = \sha1($unfurlResponse->getImageUrl());
+            $imageSaveData['width'] = $imageData[0];
+            $imageSaveData['height'] = $imageData[1];
```

Comment:
I'd merge these as an array literal, this reduces cognitive load, because the variable name only appears once:
```suggestion
            $imageSaveData = [
                'imageUrl' => $unfurlResponse->getImageUrl(),
                'imageUrlHash' => \sha1($unfurlResponse->getImageUrl()),
                'width' => $imageData[0],
                'height' => $imageData[1],
            ];
```